File: package/main/src/Array/quickSort.ts
import { compareFunctionDefault } from "./compareFunctionDefault";
import { applyInsertionSortIfNeeded } from "./sortingHelpers/applyInsertionSortIfNeeded";
import { validateRange } from "./sortingHelpers/rangeValidator";

import type { CompareFunction } from "$/array/compareFunction";

/**
 * Finds the median value among three elements in the array
 * @param array The array containing the elements
 * @param a Index of first element
 * @param b Index of second element
 * @param c Index of third element
 * @param compareFunction Function to compare elements
 * @returns The median value among the three elements
 */
const medianOfThree = <T>(
  array: T[],
  a: number,
  b: number,
  c: number,
  compareFunction: CompareFunction<T>,
): T => {
  const ab = compareFunction(array[a], array[b]);
  if (ab < 0) {
    const bc = compareFunction(array[b], array[c]);
    if (bc < 0) {
      return array[b];
    }
    if (compareFunction(array[a], array[c]) < 0) {
      return array[c];
    }
    return array[a];
  }
  const ac = compareFunction(array[a], array[c]);
  if (ac < 0) {
    return array[a];
  }
  if (compareFunction(array[b], array[c]) < 0) {
    return array[c];
  }
  return array[b];
};

/**
 * Partitions the array around a pivot element using median-of-three strategy
 * @param array Array to partition
 * @param low Starting index of the partition range
 * @param high Ending index of the partition range
 * @param compareFunction Function to compare elements
 * @returns Index of the partition point
 */
const partition = <T>(
  array: T[],
  low: number,
  high: number,
  compareFunction: CompareFunction<T>,
): number => {
  const pivot = medianOfThree(
    array,
    low,
    Math.floor((low + high) / 2),
    high,
    compareFunction,
  );
  let left = low - 1;
  let right = high + 1;

  while (true) {
    do {
      left++;
    } while (compareFunction(array[left], pivot) < 0);
    do {
      right--;
    } while (compareFunction(array[right], pivot) > 0);

    if (left >= right) {
      return right;
    }

    [array[left], array[right]] = [array[right], array[left]];
  }
};

/**
 * Internal implementation of the quicksort algorithm with tail-call optimization
 * @param array Array to sort
 * @param lowInit Initial low index of the range to sort
 * @param highInit Initial high index of the range to sort
 * @param compareFunction Function to compare elements
 * @param insertionSortThreshold Size threshold for switching to insertion sort
 */
const sortImpl = <T>(
  array: T[],
  lowInit: number,
  highInit: number,
  compareFunction: CompareFunction<T>,
  insertionSortThreshold: number,
): void => {
  let low = lowInit;
  let high = highInit;

  while (low < high) {
    if (
      applyInsertionSortIfNeeded(
        array,
        low,
        high,
        compareFunction,
        insertionSortThreshold,
      )
    ) {
      return;
    }
    const pivotIndex = partition(array, low, high, compareFunction);
    if (pivotIndex - low < high - pivotIndex) {
      sortImpl(array, low, pivotIndex, compareFunction, insertionSortThreshold);
      low = pivotIndex + 1;
    } else {
      sortImpl(
        array,
        pivotIndex + 1,
        high,
        compareFunction,
        insertionSortThreshold,
      );
      high = pivotIndex;
    }
  }
};

/**
 * Sorts an array using a hybrid algorithm combining QuickSort and InsertionSort
 * @param {T[]} array Array to sort
 * @param {CompareFunction<T>} compareFunction Comparison function that returns negative if a < b, zero if a = b, positive if a > b
 * @param {number} startIndex Starting index for the sort range (default: 0)
 * @param {number} endIndex Ending index for the sort range (default: array.length - 1)
 * @param {number} insertionSortThreshold Threshold for switching to insertion sort (default: 10)
 * @returns {T[]} Sorted array
 * @example
 * quickSort([1, 3, 2, 4, 5]); // [1, 2, 3, 4, 5]
 * quickSort([1, 3, 2], (a, b) => b - a); // [3, 2, 1]
 * quickSort(['b', 'a', 'c']); // ['a', 'b', 'c']
 */
export const quickSort = <T>(
  array: T[],
  compareFunction: CompareFunction<T> = compareFunctionDefault<T>,
  startIndex = 0,
  endIndex = array.length - 1,
  insertionSortThreshold = 10,
): T[] => {
  const result = [...array];
  const {
    startIndex: validStartIndex,
    endIndex: validEndIndex,
    shouldSort,
  } = validateRange(result, startIndex, endIndex);

  if (shouldSort) {
    sortImpl(
      result,
      validStartIndex,
      validEndIndex,
      compareFunction,
      insertionSortThreshold,
    );
  }
  return result;
};

File: package/umt_python/src/array/quick_sort.py
from typing import Callable, TypeVar

T = TypeVar("T")

CompareFunction = Callable[[T, T], int | float]


def _compare_function_default(a: T, b: T) -> int | float:
    """Default comparison function."""
    if a < b:  # type: ignore
        return -1
    if a > b:  # type: ignore
        return 1
    return 0


def _insertion_sort_range(
    array: list[T],
    start: int,
    end: int,
    compare_function: CompareFunction[T],
) -> None:
    """Sort a range using insertion sort."""
    for i in range(start + 1, end + 1):
        key = array[i]
        j = i - 1
        while j >= start and compare_function(array[j], key) > 0:
            array[j + 1] = array[j]
            j -= 1
        array[j + 1] = key


def _median_of_three(
    array: list[T],
    a: int,
    b: int,
    c: int,
    compare_function: CompareFunction[T],
) -> T:
    """Finds the median value among three elements in the array."""
    ab = compare_function(array[a], array[b])
    if ab < 0:
        bc = compare_function(array[b], array[c])
        if bc < 0:
            return array[b]
        if compare_function(array[a], array[c]) < 0:
            return array[c]
        return array[a]
    ac = compare_function(array[a], array[c])
    if ac < 0:
        return array[a]
    if compare_function(array[b], array[c]) < 0:
        return array[c]
    return array[b]


def _partition(
    array: list[T],
    low: int,
    high: int,
    compare_function: CompareFunction[T],
) -> int:
    """Partitions the array around a pivot element."""
    pivot = _median_of_three(array, low, (low + high) // 2, high, compare_function)
    left = low - 1
    right = high + 1

    while True:
        left += 1
        while compare_function(array[left], pivot) < 0:
            left += 1

        right -= 1
        while compare_function(array[right], pivot) > 0:
            right -= 1

        if left >= right:
            return right

        array[left], array[right] = array[right], array[left]


def _sort_impl(
    array: list[T],
    low_init: int,
    high_init: int,
    compare_function: CompareFunction[T],
    insertion_sort_threshold: int,
) -> None:
    """Internal implementation of the quicksort algorithm."""
    low = low_init
    high = high_init

    while low < high:
        if high - low < insertion_sort_threshold:
            _insertion_sort_range(array, low, high, compare_function)
            return

        pivot_index = _partition(array, low, high, compare_function)
        if pivot_index - low < high - pivot_index:
            _sort_impl(
                array, low, pivot_index, compare_function, insertion_sort_threshold
            )
            low = pivot_index + 1
        else:
            _sort_impl(
                array, pivot_index + 1, high, compare_function, insertion_sort_threshold
            )
            high = pivot_index


def quick_sort(
    array: list[T],
    compare_function: CompareFunction[T] | None = None,
    start_index: int = 0,
    end_index: int | None = None,
    insertion_sort_threshold: int = 10,
) -> list[T]:
    """
    Sorts an array using a hybrid algorithm combining QuickSort and InsertionSort.

    Args:
        array: Array to sort.
        compare_function: Comparison function that returns negative if a < b,
                          zero if a = b, positive if a > b.
        start_index: Starting index for the sort range (default 0).
        end_index: Ending index for the sort range (default array.length - 1).
        insertion_sort_threshold: Threshold for switching to insertion sort (default 10).

    Returns:
        Sorted array.

    Example:
        >>> quick_sort([1, 3, 2, 4, 5])
        [1, 2, 3, 4, 5]
        >>> quick_sort([1, 3, 2], lambda a, b: b - a)
        [3, 2, 1]
        >>> quick_sort(['b', 'a', 'c'])
        ['a', 'b', 'c']
    """
    result = list(array)
    if len(result) <= 1:
        return result

    if compare_function is None:
        compare_function = _compare_function_default

    if end_index is None:
        end_index = len(result) - 1

    valid_start = max(0, min(start_index, len(result) - 1))
    valid_end = max(0, min(end_index, len(result) - 1))

    if valid_start < valid_end:
        _sort_impl(
            result, valid_start, valid_end, compare_function, insertion_sort_threshold
        )

    return result
